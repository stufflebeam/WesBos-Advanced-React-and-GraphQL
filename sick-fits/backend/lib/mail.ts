import { createTransport, getTestMessageUrl } from 'nodemailer';

const transport = createTransport({
    host: process.env.ETHEREAL_MAIL_HOST || process.env.MAIL_HOST,
    port: process.env.ETHEREAL_MAIL_PORT || process.env.MAIL_PORT,
    auth: {
        user: process.env.ETHEREAL_MAIL_USER || process.env.MAIL_USER,
        pass: process.env.ETHEREAL_MAIL_PASS || process.env.MAIL_PASS,
    },
});

const frontEndURL =
    process.env.SICKFITS_FRONTEND_URL || process.env.FRONTEND_URL;

function makeANiceEmail(text: string): string {
    return `
    <div className="email" style="
        border: 1px solid black;
        padding: 20px;
        font-family: sans-serif;
        line-height: 2;
        font-size: 20px;
    ">
        <h2>Hello There!</h2>
        <p>${text}</p>

        <p>ðŸ˜˜, Sickfits</p>
    </div>
    `;
}

export interface MailResponse {
    accepted?: string[] | null;
    rejected?: null[] | null;
    envelopeTime: number;
    messageTime: number;
    messageSize: number;
    response: string;
    envelope: Envelope;
    messageId: string;
}
export interface Envelope {
    from: string;
    to?: string[] | null;
}

export async function sendPasswordResetEmail(
    resetToken: string,
    to: string
): Promise<void> {
    // email the user a token/link to reset their password
    const info = (await transport.sendMail({
        to,
        from: 'Sickfits <test@example.com>',
        subject: 'Reset your password on Sickfits',
        html: makeANiceEmail(`Please use the following link to reset your password:<br/>
        <a href="${frontEndURL}/reset?token=${resetToken}">Reset Password</a><br/>
        <br/>
        If, for some reason, you cannot click the link, copy and paste the following into your browser:<br/>
        ${frontEndURL}/reset?token=${resetToken} <br/>
        If you did not request this, please ignore this email and your password will remain unchanged.
        `),
    })) as MailResponse;

    // console.log('[mail] info:', info);
    // console.log('[mail] Message sent: %s', info.messageId);

    // Of course, being that we are currently using an environment variable called
    // ETHEREAL_MAIL_USER, this conditional is not really necessary. But, it's good
    // to have it in here for when I build in more robust environment switching.
    if (process.env.ETHEREAL_MAIL_USER.includes('ethereal.email')) {
        console.log(
            '[mail] ðŸ“§ Message Sent! Preview URL: %s',
            getTestMessageUrl(info)
        );
    }
}
