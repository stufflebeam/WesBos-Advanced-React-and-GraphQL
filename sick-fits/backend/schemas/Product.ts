import { list } from '@keystone-next/keystone/schema';
import { integer, relationship, select, text } from '@keystone-next/fields';
import { cloudinaryImage } from '@keystone-next/cloudinary';
import { isSignedIn, permissions } from '../access';

export const Product = list({
  access: {
    create: isSignedIn,
    read: permissions.isSuperAdmin,
    update: isSignedIn,
    delete: isSignedIn,
  },
  ui: {
    listView: {
      initialColumns: ['name', 'price', 'status'],
    },
  },
  fields: {
    name: text({ isRequired: true }),
    description: text({
      ui: {
        displayMode: 'textarea',
      },
    }),
    photo: relationship({
      ref: 'ProductImage.product',
      ui: {
        displayMode: 'cards',
        cardFields: ['image', 'altText'],
        inlineCreate: { fields: ['image', 'altText'] },
        inlineEdit: { fields: ['image', 'altText'] },
      },
    }),
    status: select({
      options: [
        { label: 'Draft', value: 'DRAFT' },
        { label: 'Available', value: 'AVAILABLE' },
        { label: 'Unavailable', value: 'UNAVAILABLE' },
      ],
      defaultValue: 'DRAFT',
      ui: {
        displayMode: 'segmented-control',
        // createView: { fieldMode: 'hidden' },
        // editView: { fieldMode: 'hidden' },
      },
    }),
    price: integer(),
  },
});
