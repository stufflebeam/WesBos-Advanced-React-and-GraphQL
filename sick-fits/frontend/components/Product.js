import Link from 'next/link';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPencil, faTrash } from '@fortawesome/free-solid-svg-icons';
import formatMoney from '../lib/formatMoney';
import ItemStyles from './styles/ItemStyles';
import PriceTag from './styles/PriceTag';
import Title from './styles/Title';
import DeleteProduct from './DeleteProduct';
import AddToCart from './AddToCart';

// This is one example of a placeholder image that can be used when a product
// doesn't have a photo.
// TODO: Ideally, we would use different placeholder images for different products.
// For example, we could use a shoe placeholder image for shoes, a hat placeholder
// image for hats, and so on.
// This would probably be easiest if we added a 'category' field to each of the products.
// If a product can be in multiple categories, we could designate one as the 'primary'
// or just use the first one. Not that important at the moment, but it's something to
// keep in mind, as having category-tailored placeholder images would add to the
// fit and finish of the site.
// For now though, we'll just use the same placeholder image for all products.
const placeholder =
  'https://res.cloudinary.com/stufflebeam/image/upload/v1661582104/placeholder/placeholder-images-product-2_large.webp';

// TODO: Once the product grid is fully styled, if there is still an issue with long
//       descriptions looking terrible, then we should truncate them at word boundaries
//       and add an ellipsis.
//       We can wire that up ourselves, but it'd be quickest and easiest to just use
//       something like this:
//       https://www.npmjs.com/package/ellipsize
//       If we end up supporting HTML in descriptions, then we'll need to use something
//       like this instead:
//       https://www.npmjs.com/package/react-truncate-html
//       It might also be better to just return the first line of the description
//       or add a 'short description' field or a 'subhead' field to the products table.

// TODO: Conditionally render the Buy and Edit buttons, depending on whether or not
//       the user is signed in and whether or not they have the correct permissions.

// TODO: Conditionally render the products in the grid, depending on whether or not
//       the products are available or not.

// const Product = ({ product }) => (
export default function Product({ product }) {
  return (
    <ItemStyles key={product.id}>
      <img
        src={product?.photo?.image?.publicUrlTransformed || placeholder}
        alt={product?.photo?.altText || product?.name}
      />
      <Title>
        <Link href={`/product/${product.id}`}>{product.name}</Link>
      </Title>
      <PriceTag>{formatMoney(product.price)}</PriceTag>
      <p>{product.description}</p>
      <div className="buttonList">
        {/* <Link href={`/product/${product.id}`}>
          <a>View Product</a>
        </Link> */}
        <Link
          href={{
            pathname: '/update',
            query: { id: product.id },
          }}
        >
          {/* Have to wrap in an anchor tag because Link expects a single chile */}
          <button type="button">
            {/* TODO: Add custom styling to these Font Awesome icons to give them a bit of margin */}
            Edit <FontAwesomeIcon icon={faPencil} className="fa-icon" />
          </button>
        </Link>
        <DeleteProduct id={product.id}>
          Delete <FontAwesomeIcon icon={faTrash} className="fa-icon" />
        </DeleteProduct>
        <AddToCart id={product.id} />
      </div>
    </ItemStyles>
  );
}

// export default Product;
