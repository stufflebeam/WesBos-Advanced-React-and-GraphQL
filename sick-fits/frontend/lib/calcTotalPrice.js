export default function calcTotalPrice(cart) {
  //   console.log('[calcTotalPrice] cart', cart);
  if (!cart) return 0;
  return cart.reduce((tally, cartItem) => {
    // We run the following check because it is possible for an item to
    // be deleted while it is still in a user's cart.
    if (!cartItem.product) return tally;
    return tally + cartItem.product.price * cartItem.quantity;
  }, 0);
}
