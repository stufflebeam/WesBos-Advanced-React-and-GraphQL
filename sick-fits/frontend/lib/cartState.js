import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  // This is our own custom provider
  // We will store data (state) and functionality (updaters) here
  // and anyone can access this data and the updaters via the consumer
  //   const cartOpen = true;
  const [cartOpen, setCartOpen] = useState(true);

  function toggleCart() {
    setCartOpen(!cartOpen);
  }

  function openCart() {
    setCartOpen(true);
  }

  function closeCart() {
    setCartOpen(false);
  }

  return (
    <LocalStateProvider
      value={{ cartOpen, setCartOpen, toggleCart, openCart, closeCart }}
    >
      {children}
    </LocalStateProvider>
  );
}

// This is a custom hook for accessing the data in our provider (the cart local state)
function useCart() {
  // we use a consumer here, to access the local state
  const all = useContext(LocalStateContext);
  return all;
}

export { CartStateProvider, LocalStateContext, useCart };
