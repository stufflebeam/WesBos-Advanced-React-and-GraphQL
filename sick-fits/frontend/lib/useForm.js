import { useEffect, useState } from 'react';

export default function useForm(initial = {}) {
  // create a state object for our form inputs
  const [inputs, setInputs] = useState(initial);
  // const initialValues = Object.values(initial).join('');
  // This 'feels' better than the suggested solution (above), as it avoids useEffect()
  // being called due to object property order not being guaranteed.
  // This way, it only gets called when the form is first populated with fetched data
  // and the values go from '' to something else.
  const initialValuesEmpty = Object.values(initial).join('') === '';
  // console.log('[useForm] initialValuesEmpty: ', initialValuesEmpty);

  useEffect(() => {
    // This function runs when the things we are watching change
    // console.log(
    //   '[useForm] useEffect() initialValuesEmpty: ',
    //   initialValuesEmpty
    // );
    setInputs(initial);
  }, [initialValuesEmpty]);

  // create a function that updates the state object
  const handleChange = (e) => {
    let { value, name, type } = e.target;

    // check to see if the input is a number, and if it is, convert it to a number
    if (type === 'number') {
      value = parseInt(value);
    }

    // check to see if the input is a file and handle it accordingly if so
    if (type === 'file') {
      [value] = e.target.files;
    }

    // console.log('[handleChange]', { value, name, type });
    setInputs({
      // copy the existing state first
      ...inputs,
      // update the state with the new input value
      [name]: value,
    });
  };

  function resetForm() {
    setInputs(initial);
  }

  function clearForm() {
    // setInputs({}); // doesn't work
    const blankState = Object.fromEntries(
      Object.entries(inputs).map(([key, value]) => [key, ''])
    );
    setInputs(blankState);
  }

  // return the things we wan to surface from this custom hook
  // specifically, an object that includes the state and a function to update it
  return { inputs, handleChange, resetForm, clearForm };
}
