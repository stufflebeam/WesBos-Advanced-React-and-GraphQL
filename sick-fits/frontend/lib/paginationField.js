// import {PAGINATION_QUERY} from '../graphql/queries'; // TODO: move all queries to graphql folder
import { PAGINATION_QUERY } from '../components/Pagination';

export default function paginationField() {
  return {
    keyArgs: false, // This tells Apollo that we will take care of everything.
    read(existing = [], { args, cache }) {
      //   console.log({ existing, args, cache });
      const { skip, first } = args;
      const data = cache.readQuery({ query: PAGINATION_QUERY });
      //   console.log({ first, skip, data });
      const count = data?._allProductsMeta?.count;
      const page = skip / first + 1; // Math.ceil(skip / first) + 1;
      const pageCount = Math.ceil(count / first);
      console.log('[paginationField] read():', {
        skip,
        first,
        count,
        page,
        pageCount,
      });

      // Check to see if we have the requested items in the cache.
      const items = existing.slice(skip, skip + first).filter((x) => x); // TODO: Should this be .filter((x) => !!x);

      /* 
      IF
        There are items
        AND the number of items is not enough to satisfy the request (i.e. there are not enough to fill the page)
        AND we are on the last page
      THEN just send the data back to Apollo
      */

      // TODO: Handle the case where a user tries to navigate to a page that doesn't exist.
      //       For example, if there are 10 items total, and we're showing 4 items per page,
      //       and the user navigates to page 5, we should either route them to the main
      //       products page OR route them to the last products page by updating the URL
      //       from products/5 to products/3 and showing them the last 2 items.

      if (items.length && items.length !== first && page === pageCount) {
        return items;
      }

      if (items.length !== first) {
        // If we don't have all the items, we need to fetch them.
        return false;
      }

      // If we have all the items, we can just return them.
      if (items.length) {
        console.log(
          `[paginationField]: There are ${items.length} items in the cache. Sending them to Apollo...`
        );
        return items;
      }

      // If either of the previous return statements didn't return anything, we need to fetch the items,
      // so we return false.
      return false;

      // When we run a query, the first thing that Apollo does is ask the read() function for those items.
      // When this happens, we can do one of two things:
      // 1. If we have the data, we can return it.
      // 2. If we don't have the data, we can return null or false, which will tell Apollo to run the query via the network.
    },
    merge(existing, incoming, { args, cache }) {
      // When Apollo gets query results back from the network for items that the read() function returned false for,
      // it will call the merge() function to merge the results into the cache.
      // So, this is where we decide how we want to put the returned data into the cache.
      const { skip, first } = args;

      console.log('[paginationField]: merging items from the network...]', {
        existing,
        incoming,
        args,
        cache,
      });
      const merged = existing ? existing.slice(0) : [];

      for (let i = skip; i < skip + incoming.length; ++i) {
        merged[i] = incoming[i - skip];
      }
      console.log('[paginationField] merge():', { merged });

      // return the merged items
      // at this point, Apollo will call the read() function again to see if it can return the
      // merged items from the cache.
      return merged;
    },
  };
}
