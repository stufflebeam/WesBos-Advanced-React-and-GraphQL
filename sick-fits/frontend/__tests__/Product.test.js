// import { render } from "../node_modules/@testing-library/react/types/index";
import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import Product from '../components/Product';
import { fakeItem } from '../lib/testUtils';
import { CartStateProvider } from '../lib/cartState';

const product = fakeItem();

describe('<Product/>', () => {
  it('renders out the price tag and title', () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider>
          <Product product={product} />
        </MockedProvider>
      </CartStateProvider>
    );
    // debug();
    const priceTag = screen.getByText('$50');
    // debug(priceTag);
    expect(priceTag).toBeInTheDocument();
    const link = container.querySelector('a');
    debug(link);
    // These values can be explicitly set here or be pulled in dynamically from the mock data
    expect(link).toHaveAttribute('href', '/product/abc123');
    expect(link).toHaveAttribute('href', `/product/${product.id}`);
    expect(link).toHaveTextContent(product.name);
  });

  it('renders and matches the snapshot', () => {
    const { container, debug } = render(
      <CartStateProvider>
        <MockedProvider>
          <Product product={product} />
        </MockedProvider>
      </CartStateProvider>
    );

    // TODO: Determine how snapshot testing works with dynamically-generated classes
    //       like those created by styled-components.
    expect(container).toMatchSnapshot();
  });
});
