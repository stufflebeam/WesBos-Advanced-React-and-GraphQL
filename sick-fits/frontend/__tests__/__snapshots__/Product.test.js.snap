// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product/> renders and matches the snapshot 1`] = `
<div>
  <div
    class="ItemStyles-sc-16pk14u-0 ifLwZc"
  >
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <h3
      class="Title-sc-16nq74k-0 fPQljb"
    >
      <a
        href="/product/abc123"
      >
        dogs are best
      </a>
    </h3>
    <span
      class="PriceTag-sc-nwbk6t-0 ffUdgI"
    >
      $50
    </span>
    <p>
      dogs
    </p>
    <div
      class="buttonList"
    >
      <button
        type="button"
      >
        Edit 
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-pencil fa-icon"
          data-icon="pencil"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M421.7 220.3L188.5 453.4L154.6 419.5L158.1 416H112C103.2 416 96 408.8 96 400V353.9L92.51 357.4C87.78 362.2 84.31 368 82.42 374.4L59.44 452.6L137.6 429.6C143.1 427.7 149.8 424.2 154.6 419.5L188.5 453.4C178.1 463.8 165.2 471.5 151.1 475.6L30.77 511C22.35 513.5 13.24 511.2 7.03 504.1C.8198 498.8-1.502 489.7 .976 481.2L36.37 360.9C40.53 346.8 48.16 333.9 58.57 323.5L291.7 90.34L421.7 220.3zM492.7 58.75C517.7 83.74 517.7 124.3 492.7 149.3L444.3 197.7L314.3 67.72L362.7 19.32C387.7-5.678 428.3-5.678 453.3 19.32L492.7 58.75z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        type="button"
      >
        Delete 
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-trash fa-icon"
          data-icon="trash"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M135.2 17.69C140.6 6.848 151.7 0 163.8 0H284.2C296.3 0 307.4 6.848 312.8 17.69L320 32H416C433.7 32 448 46.33 448 64C448 81.67 433.7 96 416 96H32C14.33 96 0 81.67 0 64C0 46.33 14.33 32 32 32H128L135.2 17.69zM394.8 466.1C393.2 492.3 372.3 512 346.9 512H101.1C75.75 512 54.77 492.3 53.19 466.1L31.1 128H416L394.8 466.1z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        type="button"
      >
        Add
         to Cart!
         
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-cart-plus fa-icon"
          data-icon="cart-plus"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M96 0C107.5 0 117.4 8.19 119.6 19.51L121.1 32H541.8C562.1 32 578.3 52.25 572.6 72.66L518.6 264.7C514.7 278.5 502.1 288 487.8 288H170.7L179.9 336H488C501.3 336 512 346.7 512 360C512 373.3 501.3 384 488 384H159.1C148.5 384 138.6 375.8 136.4 364.5L76.14 48H24C10.75 48 0 37.25 0 24C0 10.75 10.75 0 24 0H96zM272 180H316V224C316 235 324.1 244 336 244C347 244 356 235 356 224V180H400C411 180 420 171 420 160C420 148.1 411 140 400 140H356V96C356 84.95 347 76 336 76C324.1 76 316 84.95 316 96V140H272C260.1 140 252 148.1 252 160C252 171 260.1 180 272 180zM128 464C128 437.5 149.5 416 176 416C202.5 416 224 437.5 224 464C224 490.5 202.5 512 176 512C149.5 512 128 490.5 128 464zM512 464C512 490.5 490.5 512 464 512C437.5 512 416 490.5 416 464C416 437.5 437.5 416 464 416C490.5 416 512 437.5 512 464z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
